### DISCUSSION TITLE
Using `times` syscall for cross-target library builds in `clock` function

### DISCUSSION THREAD

**Discussion by M-Moawad** [picolibc/picolibc#710] [Q&A]
Greetings!

I have a question according the current behavior of clock function.

Clock function, here https://github.com/picolibc/picolibc/blob/main/newlib/libc/time/clock.c, now uses `times` syscall to calculate the CPU time consumed by the process.
```c
clock_t 
clock (void)
{
  struct tms tim_s;
  clock_t res;

  if ((res = (clock_t) times (&tim_s)) != (clock_t) -1)
    res = (clock_t) (tim_s.tms_utime + tim_s.tms_stime +
		     tim_s.tms_cutime + tim_s.tms_cstime);

  return res;
}
```
When this code is executed on a simulator on the host machine, the clock used for measuring time will typically be the **host machine's clock**, not the simulated **target machine's clock**, this is because the `times` function is a system call that interacts with the host machine's operating system to retrieve timing information about the process.

The question here: Is this behavior correct? Or is it needed to have a target dependent clock function to use the target's clock?  

Any help is much appreciated in advance!

**Comment by keith-packard** [comment]
I'm not sure what you want here -- if you have a cycle-accurate simulator, then perhaps that should be computing time based upon the emulated cycle clock. However, many simulators are behavioral and don't accurately reflect how long operations would take on the emulated machine. In that case, it seems like the host time is about as good as you can get?

**Reply by abrodkin** [reply]
Well, it's not only about emulated targets, I'd say it's even more relevan in case of real hardware targets (even if it's implemented as an FPGA system or even RTL-simulation but not an off-the-shelf board). And the point is on those targets it's usually possible to get very precise time: either from a counter which is incremented with the same frequency as the CPU clock signal or at least from some RTC (Real-Time Clock) which is still very precise. In that sense request of time from the host via say JTAG interface will give pretty low precision because there's at least a communication overhead.

And so, the question is how to use precise clock source of the target instead of using semihosting obtained host time. And do it in a graceful manner. Clearly that's a platform (sometime even architecture) specific feature as all platforms are different.

The first think which comes to my mind is introduction of yet another Picolibc option like `PICOLIBC_USE_TARGET_CLOCK` and then use platform-specific implementation of `clock()`. Apparently LLVM provides a very nice `__builtin_readcyclecounter()` (see https://clang.llvm.org/docs/LanguageExtensions.html#builtin-readcyclecounter) which could be used on some targets, namely AArch64, ARM, ARC (in our downstream compiler as upstream port is not yet available), PowerPC and RISC-V. So that could be used as a nice "generic" implementation. Note though, while on some architectures there're some conditions (ARM requires performance monitor extension to be available in the design, ARC requires at least one of core timers to be configured in the design), on RISC-V it always works just because `cycle` CSR is a part of the base RV32 or RV64 configuration.

Does that look OK to you?

**Reply by keith-packard** [reply]
The `times` interface doesn't use any time reference, it's effectively a fancier version of the `clock` function, splitting the losw-level clock ticks it reports among four buckets. It sounds like you just want to replace the `clock` function which uses `times` in its implementation. There's no problem doing that in a picolibc-using project, you can just provide a clock implementation in your application and the one in picolibc won't be used.

You could also provide an implementation of the `times` interface using the primitive clocks.

I guess I'm not clear on what functionality within picolibc you're trying to modify and not simply replace?

**Reply by abrodkin** [reply]
Surely it's possible to replace either `times()` or `clock()` outside of the Picolibc, but the idea is why don't we make Picolibc out of the box usable for making precise measurements. Think of say `dhrystone` which we'll be able to compile w/o any changes and still get a meaningful & as precise as possible measurements. Essentially, it all in assumption there's an implementation which works for wide range of platforms.

**Reply by keith-packard** [reply]
I must be missing something -- both `times` and `clock` are defined to return values in microseconds by POSIX (C doesn't define `CLOCKS_PER_SEC` as far as I can tell). The last time I tried compiling Dhrystone, I had to hack the code to keep it from using a clock based in seconds, but I don't recall why.
